hazelcast/src/main/java/com/hazelcast/sql/impl/client/SqlClientService.java

```diff
@@ -0,0 +1,218 @@
+                query.getTimeoutMillis(),
+                query.getCursorBufferSize()
+            );
+
+            SqlExecuteResponse response = invoke(message, connection, EXECUTE_DECODER);
+
+            handleResponseError(response.getError());
+
+            return new SqlClientResult(
+                this,
+                connection,
+                response.getQueryId(),
+                response.getRowMetadata(),
+                response.getPage(),
+                query.getCursorBufferSize()
+            );
+        } catch (Exception e) {
+            throw rethrow(e, connection);
+        }
+    }
+
+    /**
+     * Fetch the next page of the given query.
+     *
+     * @param connection Connection.
+     * @param queryId Query ID.
+     * @return Pair: fetched rows + last page flag.
+     */
+    public SqlPage fetch(Connection connection, QueryId queryId, int cursorBufferSize) {
+        try {
+            ClientMessage message = SqlFetchCodec.encodeRequest(queryId, cursorBufferSize);
+
+            SqlFetchResponse res = invoke(message, connection, FETCH_DECODER);
+
+            handleResponseError(res.getError());
+
+            return res.getPage();
+        } catch (Exception e) {
+            throw rethrow(e, connection);
+        }
+    }
+
+    /**
+     * Close remote query cursor.
+     *
+     * @param connection Connection.
+     * @param queryId Query ID.
+     */
+    void close(Connection connection, QueryId queryId) {
+        try {
+            ClientMessage request = SqlCloseCodec.encodeRequest(queryId);
+
+            invoke(request, connection, CLOSE_DECODER);
+        } catch (Exception e) {
+            throw rethrow(e, connection);
+        }
+    }
+
+    private Data serializeParameter(Object parameter) {
+        try {
+            return getSerializationService().toData(parameter);
+        } catch (Exception e) {
+            throw rethrow(
+                QueryException.error("Failed to serialize query parameter " + parameter + ": " + e.getMessage())
+            );
+        }
+    }
+
+    Row deserializeRow(Data data) {
+        try {
+            return getSerializationService().toObject(data);
+        } catch (Exception e) {
+            throw rethrow(
+                QueryException.error("Failed to deserialize query result row: " + e.getMessage())
+            );
+        }
+    }
+
+    private UUID getClientId() {
+        return client.getLocalEndpoint().getUuid();
+    }
+
+    private InternalSerializationService getSerializationService() {
+        return client.getSerializationService();
+    }
+
+    private <T> T invoke(ClientMessage request, Connection connection, ClientMessageDecoder<T> decoder) throws Exception {
+        ClientInvocation invocation = new ClientInvocation(client, request, null, connection);
+
+        ClientInvocationFuture fut = invocation.invoke();
+
+        return new ClientDelegatingFuture<T>(fut, getSerializationService(), decoder, false).get();
```

Comment:
You don't need the ClientDelegatingFuture since you are doing a sync call. It is mostly needed when we need to pass a future to the user(or other part of the system)

Why not just do the following instead? 
```
ClientMessage clientMessage = fut.get();
return decoder.decodeClientMessage(clientMessage);
```

Even further, there is no need for static decoder functions because they are all sync calls. 
I would suggest making the return type `ClientMessage`.  For the fetch method: 
```
            ClientMessage responseMessage = invoke(message, connection);
            SqlFetchCodec.ResponseParameters response = SqlFetchCodec.decodeResponse(responseMessage);
            SqlFetchResponse res = new SqlFetchResponse(response.rowPage, response.error);
            handleResponseError(res.getError());
```

Static decoder functions and delegating future are used for async cases. There is no need to complicate the code with them when not needed.

Reply:
Simplified as you suggested.